### PR TITLE
Added tests for trace callable and runnable

### PR DESCRIPTION
**Pull request by marcingrzejszczak**
@spencergibb @dsyer - can you please take a look at this and tell me if also in your opinion those tests should pass? :)

### CODE REVIEW

File: spring-cloud-sleuth-core/src/test/java/org/springframework/cloud/sleuth/instrument/TraceCallableTest.java [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+	TraceManager traceManager = Mockito.mock(TraceManager.class);
+
+	@Test
+	@Ignore("Will fail because trace is not removed after callable gets executed")
+	public void should_remove_span_from_thread_local_after_finishing_work() throws Exception {
+		givenCallableGetsSubmitted(thatSetsTraceInCurrentThreadLocalWithInitialTrace());
+
+		Trace secondTrace = whenCallableGetsSubmitted(thatRetrievesTraceFromThreadLocal());
+
+		then(secondTrace).isNull();
```
Comment (dsyer) [inline spring-cloud-sleuth-core/src/test/java/org/springframework/cloud/sleuth/instrument/TraceCallableTest.java]: I don't think this test should pass.

File: spring-cloud-sleuth-core/src/test/java/org/springframework/cloud/sleuth/instrument/TraceCallableTest.java [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+	}
+
+	private Callable<Trace> thatSetsTraceInCurrentThreadLocalWithInitialTrace() {
+		return new Callable<Trace>() {
+			@Override
+			public Trace call() throws Exception {
+				TraceContextHolder.setCurrentTrace(Mockito.mock(Trace.class));
```
Comment (dsyer) [inline spring-cloud-sleuth-core/src/test/java/org/springframework/cloud/sleuth/instrument/TraceCallableTest.java]: Usage of `TraceContextHolder` is basically banned or at least heavily discouraged so this should not be in any real user code.

File: spring-cloud-sleuth-core/src/test/java/org/springframework/cloud/sleuth/instrument/TraceCallableTest.java (hunk 1 above)
Comment (dsyer) [inline spring-cloud-sleuth-core/src/test/java/org/springframework/cloud/sleuth/instrument/TraceCallableTest.java]: This one should pass:

```
@RunWith(MockitoJUnitRunner.class)
public class TraceCallableTest {

	ExecutorService executor = Executors.newSingleThreadExecutor();
	TraceManager traceManager = new DefaultTraceManager(new AlwaysSampler(),
			new RandomUuidGenerator(), Mockito.mock(ApplicationEventPublisher.class));

	@Test
	public void should_remove_span_from_thread_local_after_finishing_work()
			throws Exception {
		Trace firstTrace = givenCallableGetsSubmitted(thatSetsTraceInCurrentThreadLocalWithInitialTrace());

		Trace secondTrace = whenCallableGetsSubmitted(thatRetrievesTraceFromThreadLocal());

		then(secondTrace.getSpan().getTraceId()).isNotEqualTo(firstTrace.getSpan().getTraceId());
	}

	private Callable<Trace> thatSetsTraceInCurrentThreadLocalWithInitialTrace() {
		return new Callable<Trace>() {
			@Override
			public Trace call() throws Exception {
				return TraceContextHolder.getCurrentTrace();
			}
		};
	}

	private Callable<Trace> thatRetrievesTraceFromThreadLocal() {
		return new Callable<Trace>() {
			@Override
			public Trace call() throws Exception {
				return TraceContextHolder.getCurrentTrace();
			}
		};
	}

	private Trace givenCallableGetsSubmitted(Callable<Trace> callable)
			throws InterruptedException, java.util.concurrent.ExecutionException {
		return whenCallableGetsSubmitted(callable);
	}

	private Trace whenCallableGetsSubmitted(Callable<Trace> callable)
			throws InterruptedException, java.util.concurrent.ExecutionException {
		return this.executor.submit(new TraceCallable<>(this.traceManager, callable))
				.get();
	}

}
```

**Comment by dsyer** [comment]
I merged this by mistake. If you want we can still discuss the expected behaviour here.

**Comment by marcingrzejszczak** [comment]
The problem emerges especially when using Hystrix commands. If we're using shared thread pools then there is an issue that a Callable / Runnable is used within a thread that has already processed a Callable / Runnable. That means that in a ThreadLocal there has already been a Span and a new Span is being processed by that Callable / Runnable. In that situation an exception is thrown because there should be nothing there set in the ThreadLocal. 

That's why I wanted to clear the entries inside the ThreadLocal after a Callable / Runnable processing took place.

Another issue is that the Hystrix command is not storing the current Trace from the current thread like it should (https://github.com/4finance/micro-infra-spring/blob/master/micro-infra-spring-base/src/main/java/com/ofg/infrastructure/hystrix/CorrelatedCommand.java#L18). But this is a topic for another PR actually...

**Comment by dsyer** [comment]
Probably the same issue being discussed in #61? I'll change the test and see if it helps flush out the stale trace.

**Comment by marcingrzejszczak** [comment]
Yup it's exactly the same thing.

When I was doing testing the app blew up in places with common threadpools
and tracedelegates.

**Pull request merged**